### PR TITLE
Page Layouts: Toggle layout assets

### LINKED ISSUE #6130: Ability to toggle layout assets
## Description 
Currently on the page layouts tab the layouts, which are individual pages from templates, have their image assets getting replaced with a single grid asset. 
This ticket is to create a toggle button (pending new design system implemented in editor) to give users the ability to see the layouts either _with_ the given assets of a page as it would a template or with the placeholder grid in its stead. 

https://www.figma.com/file/NHHMwIfxMnz39NxqkrFb7R/Stories-Post-Stable?node-id=5101%3A456645

## Acceptance Criteria 
- Whatever view the user has turned on, when the layout is selected the current page will be replaced with that layout with the proper assets for the view the user has toggled to. If a user has the placeholder grid visible in layouts the page will be rendered in the canvas with the placeholder grid. If a user has the traditional assets visible in the layouts the page will be rendered with those image assets. 
- All animations should be preserved the same as they currently are on page replacement 
- Panel should be keyboard navigable without focus trapping. 
- Toggle button should be accessible with proper labeling for screen readers. 
- Also adds a label to the panel section above the grid. 

Screen Shot 2021-02-01 at 10 14 03 AM

### PR DESCRIPTION
**Pull request by samwhale**
## Context

Want to allow users to apply page layouts _with_ and _without_ the images included in the page layouts.

[Designs:](https://www.figma.com/file/NHHMwIfxMnz39NxqkrFb7R/Stories-Post-Stable?node-id=5101%3A456645)

## Summary

- Adds new UI as show in designs above
- Clicking the toggle will add/remove image content from layouts
- Clicking the page layout will apply the layout with or without the image content

Fixed an accessibility concern in the text panes tab. Can be moved to another pr if we want it separate.

## Relevant Technical Choices

n/a

## To-do

none

## User-facing changes

|Before|After|
|--|--|
|before|after|
|before-rtl|after-rtl|

## Testing Instructions

1. Create a new story.
2. Click on the page layouts tab.
3. Click on a layout. Should add layout without the image content.
4. Clicking the toggle and the toggle label should change the toggle value.  Pressing `spacebar` when the toggle is focused should work too. (Note: there's some lag here as we re-create all the page layouts)
5. Click a page layout. Should add layout without the image content.

Things that should still work:
- changing a page layout when there is something on the canvas should show a dialog prompting you to accept deleting any existing stuff
- animations should still be preserved when the layout is applied
- Panel should still be keyboard accessible

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Create a new story.
2. Click on the page layouts tab.
3. Click on a layout. Should add layout without the image content.
4. Clicking the toggle and the toggle label should change the toggle value.  Pressing `spacebar` when the toggle is focused should work too. (Note: there's some lag here as we re-create all the page layouts)
5. Click a page layout. Should add layout without the image content.

Things that should still work:
- changing a page layout when there is something on the canvas should show a dialog prompting you to accept deleting any existing stuff
- animations should still be preserved when the layout is applied
- Panel should still be keyboard accessible

### UAT

- [x] UAT should use the same steps as above.

### Does this PR have a security-related impact?

no

### Does this PR change what data or activity we track or use?

no

### Does this PR have a legal-related impact?

no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6130

### CODE REVIEW

File: assets/src/edit-story/components/library/panes/text/textSets/textSetsPane.js [hunk 1]
```diff
@@ -154,7 +156,11 @@ function TextSetsPane({ paneRef }) {
           {PANE_TEXT.TITLE}
         </Headline>
         <TextSetsToggle>
-          <Text size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.SMALL}>
+          <Text
+            as="label"
+            htmlFor={toggleId}
+            size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.SMALL}
+          >
```
Comment (samwhale) [inline assets/src/edit-story/components/library/panes/text/textSets/textSetsPane.js:163]: Realized the toggle in the text sets pane didn't have a `label` associated with the toggle. This can be moved to another pr if we'd like as well. There are no visual changes here other than making the cursor a `pointer` when hovering the text.

File: assets/src/edit-story/components/library/panes/text/textSets/textSetsPane.js [hunk 2]
```diff
@@ -163,7 +169,6 @@ function TextSetsPane({ paneRef }) {
             name={toggleId}
             checked={showInUse}
             onChange={onChangeShowInUse}
-            label={PANE_TEXT.SWITCH_LABEL}
```
Comment (samwhale) [inline assets/src/edit-story/components/library/panes/text/textSets/textSetsPane.js:166]: `label` isn't a property that the toggle uses.

File: assets/src/edit-story/components/library/panes/pageLayouts/pageLayouts.js [hunk 3]
```diff
@@ -159,6 +184,30 @@ function PageLayouts({ pages, parentRef }) {
...
+          <Text
+            as="label"
+            htmlFor={toggleId}
+            size={THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.SMALL}
+          >
+            {__('Show Images', 'web-stories')}
+          </Text>
+          <Toggle
+            id={toggleId}
+            aria-label={__('Show images in layouts', 'web-stories')}
```
Comment (samwhale) [inline assets/src/edit-story/components/library/panes/pageLayouts/pageLayouts.js]: Do we like this aria label for the toggle?

File: assets/src/edit-story/components/library/panes/pageLayouts/pageLayoutsPane.js [hunk 4]
```diff
@@ -46,7 +46,8 @@ export const PaneInner = styled.div`
 `;
 
 export const PageLayoutsParentContainer = styled.div`
-  ${virtualPaneContainer}
+  ${virtualPaneContainer};
+  margin-top: 18px;
```
Comment (samwhale) [inline assets/src/edit-story/components/library/panes/pageLayouts/pageLayoutsPane.js:50]: We don't need the default margin top here since we added a new row of things

File: assets/src/edit-story/app/api/getAllPageLayouts.js [hunk 5]
```diff
@@ -67,12 +67,16 @@ const replacePageImagesWithPlaceholders = (page, { assetsURL }) => {
...
+  cdnURL,
+  assetsURL,
+  showImages = true,
+}) {
   const templates = await getAllTemplates({ cdnURL });
   return templates.map((template) => ({
     ...template,
     pages: template.pages.map((page) =>
-      replacePageImagesWithPlaceholders(page, { assetsURL })
+      showImages ? page : replacePageImagesWithPlaceholders(page, { assetsURL })
```
Comment (BrittanyIRL) [inline assets/src/edit-story/app/api/getAllPageLayouts.js]: 👏

**Review by BrittanyIRL — COMMENTED** [review]
Awesome, this was way easier than I thought it might be! 

I know that the call to get page layouts is memoizing the templates already but I'm curious what you think about keeping the initial page layouts that are returned after the toggle so that further fetches aren't needed since it's static data. 

Essentially you'd have like a `layoutsWithoutImages` and a `layoutsWithImages`, if populated, no call to `getPageLayouts` is made and that reference is used instead to cut down on trips. Thoughts?

File: assets/src/edit-story/app/api/getAllPageLayouts.js [hunk 6]
```diff
@@ -67,12 +67,16 @@ const replacePageImagesWithPlaceholders = (page, { assetsURL }) => {
   };
 };
 
-export default async function getAllPageLayouts({ cdnURL, assetsURL }) {
+export default async function getAllPageLayouts({
+  cdnURL,
+  assetsURL,
+  showImages = true,
```
Comment (miina) [inline assets/src/edit-story/app/api/getAllPageLayouts.js]: Looks like the default here is `true` for `showImages` but above in `getPageLayouts` it's `false` -- is this purposefully so? Should the default always be `false` as that's the default in Figma? Thoughts?

**Review by miina — APPROVED** [review]
Looks great! 🎉 

It looks like the placeholder image has changed a bit -- it has larger squares now (if I'm seeing correctly), perhaps we could replace the placeholder image here as well?

**Comment by samwhale** [comment]
> Essentially you'd have like a `layoutsWithoutImages` and a `layoutsWithImages`, if populated, no call to `getPageLayouts` is made and that reference is used instead to cut down on trips. Thoughts?

Yeah this is a great idea, then we wouldn't have to wait for the network request each time we click the toggle. I can add that in 👍

File: assets/src/edit-story/app/api/getAllPageLayouts.js (hunk 6 above)
Comment (samwhale) [reply assets/src/edit-story/app/api/getAllPageLayouts.js]: 😬 oh nice catch. Yeah I think the default should be false! I'll fix that right up

**Comment by samwhale** [comment]
> Looks great! 🎉
> 
> It looks like the placeholder image has changed a bit -- it has larger squares now (if I'm seeing correctly), perhaps we could replace the placeholder image here as well?

I'll look into this 👍

File: assets/src/edit-story/components/library/panes/pageLayouts/pageLayouts.js (hunk 3 above)
Comment (samwhale) [reply assets/src/edit-story/components/library/panes/pageLayouts/pageLayouts.js]: Thinking about this again - it may not be needed since the toggle has a `label` that's associated with it

**Comment by samwhale** [comment]
> > Looks great! 🎉
> > It looks like the placeholder image has changed a bit -- it has larger squares now (if I'm seeing correctly), perhaps we could replace the placeholder image here as well?
> 
> I'll look into this 👍

I replaced the old placeholder with one from figma. Now this looks like:
image

**Comment by samwhale** [comment]
> > Essentially you'd have like a `layoutsWithoutImages` and a `layoutsWithImages`, if populated, no call to `getPageLayouts` is made and that reference is used instead to cut down on trips. Thoughts?
> 
> Yeah this is a great idea, then we wouldn't have to wait for the network request each time we click the toggle. I can add that in 👍

Added that in - now there's only one network request (right when we want to get page layouts). Any changes to the toggle don't re-fetch templates

File: assets/src/edit-story/app/api/apiProvider.js [hunk 7]
```diff
@@ -364,9 +370,22 @@ function APIProvider({ children }) {
...
+  const getPageLayouts = useCallback(
+    async ({ showImages = false } = {}) => {
+      // check if pageLayouts have been loaded yet
+      if (pageLayouts.current.base.length === 0) {
+        pageLayouts.current.base = await getAllTemplates({ cdnURL });
+        pageLayouts.current.withoutImages = removeImagesFromPageLayouts({
+          templates: pageLayouts.current.base,
+          assetsURL,
+          showImages,
+        });
```
Comment (BrittanyIRL) [inline assets/src/edit-story/app/api/apiProvider.js:382]: I really like this solution

**Review by BrittanyIRL — APPROVED** [review]
It's so fast now! Thank you for adding tests.

**Review by embarks — APPROVED** [review]

File: packages/e2e-tests/src/specs/editor/media/insertMediaFromDialog.js [hunk 8]
```diff
@@ -24,7 +24,9 @@ import {
 
 describe('Inserting Media from Dialog', () => {
   // Uses the existence of the element's frame element as an indicator for successful insertion.
-  it('should insert an image by clicking on it', async () => {
+  // TODO #7107
+  // eslint-disable-next-line jest/no-disabled-tests
+  it.skip('should insert an image by clicking on it', async () => {
```
Comment (samwhale) [inline packages/e2e-tests/src/specs/editor/media/insertMediaFromDialog.js]: New issue: https://github.com/google/web-stories-wp/issues/7107
